source/extensions/filters/http/cache/cache_filter.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/config/filter/http/cache/v2alpha/cache.pb.h"
+
+#include "extensions/filters/http/cache/http_cache.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+/**
+ * A filter that caches responses and attempts to satisfy requests from cache.
+ */
+class CacheFilter;
+using CacheFilterSharedPtr = std::shared_ptr<CacheFilter>;
+class CacheFilter : public Http::PassThroughFilter,
+                    public std::enable_shared_from_this<CacheFilter> {
+public:
+  // Throws ProtoValidationException if no registered HttpCacheFactory for config.name.
+  static CacheFilterSharedPtr make(const envoy::config::filter::http::cache::v2alpha::Cache& config,
+                                   const std::string& stats_prefix, Stats::Scope& scope,
+                                   TimeSource& time_source) {
+    return std::shared_ptr<CacheFilter>(new CacheFilter(config, stats_prefix, scope, time_source));
```

Comment:
std::make_shared ?

Reply:
We can't use make_shared due to the private constructor. Would it be better to make the constructor public and just trust client code not to create CacheFilter objects that aren't owned by shared_ptrs?

Reply:
Explained non-use of make_shared in comment.